dilated-file-transfer.md

```diff
@@ -0,0 +1,513 @@
+
+   XXX:: maybe just lean on "version" for now? e.g. version `2` could introduce "features"?
+
+See "Example of Protocol Expansion" below for discussion about adding new attributes (including when we might increment the `"version"` instead of adding a new `"feature"`).
+
+
+## Protocol Details
+
+See the Dilation document for details on the Dilation setup procedure.
+Once a Dilation-supporting connection is open, we will have a "control" subchannel (subchannel #0).
+Either peer can also open additional subchannels.
+
+All control-channel messages are encoded using `msgpack`.
+   --> XXX: see "message encoding" discussion
+
+Control-channel message formats are described using Python pseudo-code to illustrate the data types involved.
+They are actually an encoded `Map` with `String` keys (to use `msgpack` lingo) and values as per the pseudo-code.
+
+All control-channel messages contain a integer "kind" field describing the sort of message it is.
+(That is, `"kind": "text"` for example, not the single-byte tag used for subchannel messages)
+
+Rejected idea: Version message, because we already do version negotiation via mailbox features.
+
+Rejected idea: Offer/Answer messages via the control channel: we need to open a subchannel anyway and the subchannel-IDs are not intended to be part of the Dilation public API.
+
+
+### Control Channel Messages
+
+Each side MAY send a free-form text message at any time.
+These messages look like:
+
+```python
+class Message:
+    message: str     # unicode string
+    kind: str = "text"
+```
+
+
+### Making an Offer
+
+Either side MAY propose any number of Offers at any time after the connection is set up.
+If the other peer specified `"mode": "send"` then this peer MUST NOT make any Offers.
+If this peer specified `"mode": "receive"` then this peer MUST NOT make any Offers.
+
+To make an Offer the peer opens a subchannel.
+Recall from the Dilation specification that subchannels are _record_ pipes (not simple byte-streams).
+
+All records on the subchannel begin with a single byte indicating the kind of message.
+Any additional bytes are a kind-dependent payload.
+
+The following kinds of messages exist (as indicated by the first byte):
+* 1: msgpack-encoded `FileOffer` message
+* 2: msgpack-encoded `DirectoryOffer` message
+* 3: msgpack-encoded `OfferAccept` message
+* 4: msgpack-encoded `OfferReject` message
+* 5: raw file data bytes
+* 6: msgpack-encoded `FileAcknowledge` message
+
+All other byte values are reserved for future use and MUST NOT be used.
+
+    XXX: maybe spec [0, 128) as reserved, and [128, 255) for "experiments"?
+
+The sender that opened the new subchannel MUST immediately send one of the two kinds of offer messages.
+
+To offer a single file (with message kind `1`):
+
+```python
+class FileOffer:
+    filename: str   # filename (no path segments)
+    timestamp: int  # Unix timestamp (seconds since the epoch in GMT)
+    bytes: int      # total number of bytes in the file
```

Comment:
If `bytes` would be optional, it could really help with other use cases(like /dev/urandom or any stream), but not sure if that would necessitate some `EndOfFile` message.

Reply:
Interesting .. for such a use-case, I think maybe a new message pair would be appropriate (as e.g. I'd expect the UX for accepting "a stream" is probably quite different than for a file -- e.g. if it's "a video" you'd probably pipe it directly to `mplayer` rather than trying to save it).

I think the spec as written easily allows for such an expansion (either via `"features"` or a new protocol version). Probably a `"feature"` makes the most sense since some clients may not want to use this (even if they support the newest protocol version).

Note that a new `"feature"` or protocol-version could simply declare that `bytes` is now optional (or perhaps that `null` is valid) if that worked.

I did add a `FileAcknowledge` confirmation message, but this might need a different kind of "end" message I think .. or maybe not, depending on the use-case(s).

Reply:
I think for streams an entirely different architecture makes more sense: Like for HTML, where the client makes a request to get some bytes from the server, so that it remains in control of what it receives even for infinite data streams.

However, there are other special files that require special treatment, even if not all of them should be supported. I think the most important ones are symlinks, we definitely should support them.

Reply:
What does symlink support mean?

- follow them, and send the content
    - what if it's a link outside the directory?
- send something saying "make a link"
    - what about links outside the directory?
    - what happens on Windows?
    - what happens on iOS?
    - what happens on Android?
- ask? (sender? receiver?)
- security considerations?

Reply:
Symlinks that point within the directory being sent are generally safe and maybe follow the others instead.

Assuming two clients support symlinks it should work on macOS and Linux at least and others get the followed contents instead of a symlink?

Reply:
I don't think it's at all clear that the sender should "preserve symlinks"; this isn't "a filesystem synchronization" tool or protocol. It's a tool and protocol for sending files to other devices.

While I'm sure some users might be happy to see symlinks result in (similar) symlinks on the receiver, other users will likely be surprised if their directory full of symlinks totals "0 bytes" and doesn't actually send the content to the other side, for example.

My music is organized as such: albums are symlinks to the actual .ogg tracks. So if I select an album to send to a friend I'd want the actual .ogg data to be sent (not symlinks that will be useless to them).

I think the _least_ surprising behavior is probably what Wanas notes for the Python client: follow them and send the content.

We should probably be explicit in the specification for what to do when a directory is selected and symlinks are found, though.

Reply:
Following a symlink is a recipe for disaster, just don't (at least for sending folders — for single files, you obviously want to send the file). You can make it opt-in sender side, but then still need to consider infinite loops or accidentally sending the entire hard drive and stuff like that.

Anyways we can cross this bridge when implementing the clients, for now the question is what specification support should there be.

- At the very least, there should be the possibility to send symbolic links as part of a folder.
- Optionally, we may discuss some feature flag for clients to disclose whether their file system abstraction supports them or not. (This only shifts the burden of dealing it from receiver to sender, but maybe the sender can be more graceful about the situation?)

Reply:
Why is it a "recipe for disaster"? As you say, it's obvious for some cases.

By default, `zip` follows the symlinks and stores the content.
By default, `tar` stores symlinks.

Reply:
(All of the above said, of course client-software is ultimately in charge and can choose different defaults, behavior, etc ... so we shouldn't be too prescriptive of _client behavior_ in the spec -- i.e. recommend defaults over "MUST x")

Reply:
There is no way to send symlinks in the classic protocol, so we will not include this new feature now but consider it as an expansion.

With no way to send symlinks, clients will have to choose what makes sense for their implementation when encountering one. To be consistent with the classic protocol, that means sending the content of the symlink target with the name of the symlink.

Reply:
> If it's "obvious" that selecting some symlink in "file" mode means we should follow it and send the content, it would seem for consistency that should also be the default for directories...?

Yes, kind of. Symlinks should be resolved until the top level item that should be sent. So as I said single files should be resolved. If you are sending a folder and the path you provide is a symlink or contains symlinks, those should be resolved too. But the contents folder itself should be transferred "as-is", and with that I mean no following of symlinks.

Reply:
I still don't know what that means. Is it different that what I wrote in the document?

Reply:
No, it's fine